File: csharp/go-counting/GoCounting.cs
using System;
using System.Collections.Generic;
using System.Linq;

public enum Owner
{
    None, Black, White
}

public class GoCounting
{
    private string[] board;

    public GoCounting(string input) => board = input.Split('\n');

    public (Owner owner, IEnumerable<(int, int)> tiles) Territory((int x, int y) coord)
    {
        if(coord.x < 0 || coord.y < 0 || coord.y >= board.Length || coord.x >= board[0].Length)
            throw new ArgumentException();

        var start = OwnerFor(coord);
        if(start != Owner.None)
            return (Owner.None, Enumerable.Empty<(int, int)>());
        
        var extent = Extent(coord.x, coord.y);
        var owners = extent
            .Select(o => o.owner)
            .Where(o => o != Owner.None)
            .Distinct();
        var areEmpty = extent
            .Where(o => o.owner == Owner.None)
            .Select(o => (x: o.x, y: o.y))
            .OrderBy(o => o.x).ThenBy(o => o.y);
        return (owners.Count() == 1 ? owners.First() : Owner.None, areEmpty);
    }

    private IEnumerable<(int x, int y, Owner owner)> Extent(int x, int y)
    {
        var soFar = new Dictionary<(int x, int y), Owner> { { (x, y), Owner.None } };
        var neighbours = Adjacent(x, y);
        while(neighbours.Any())
        {
            var withOwners = neighbours.Select(o => (x: o.x, y: o.y, owner: OwnerFor(o))).ToList();
            withOwners.ForEach(o => soFar.Add((o.x, o.y), o.owner));

            neighbours = withOwners
                .Where(o => o.owner == Owner.None)
                .SelectMany(o => Adjacent(o.x, o.y))
                .Where(o => !soFar.ContainsKey((o.x, o.y)))
                .ToArray();
        }

        return soFar.Select(o => (o.Key.x, o.Key.y, o.Value));
    }

    private IEnumerable<(int x, int y)> Adjacent(int x, int y)
        => new (int dx, int dy)[] { (-1, 0), (1, 0), (0, -1), (0, 1) }
            .Select(o => (x: x + o.dx, y: y + o.dy))
            .Where(o => o.x >= 0 && o.y >= 0 && o.y < board.Length && o.x < board[0].Length);

    private Owner OwnerFor(char c) => c == ' ' ? Owner.None : c == 'W' ? Owner.White : Owner.Black;

    private Owner OwnerFor((int x, int y) coord) => OwnerFor(board[coord.y][coord.x]);

    public Dictionary<Owner, IEnumerable<(int, int)>> Territories() 
    {
        var result = new Dictionary<Owner, IEnumerable<(int, int)>>
        {
            [Owner.Black] = Array.Empty<(int, int)>(),
            [Owner.White] = Array.Empty<(int, int)>(),
            [Owner.None] = Array.Empty<(int, int)>()
        };

        var allTiles = 
            Enumerable.Range(0, board.Length).SelectMany(y =>
            Enumerable.Range(0, board[0].Length).Select(x => Territory((x, y))));
        var hasTerritory = 
            allTiles.GroupBy(o => o.owner)
            .ToDictionary(o => o.Key, 
                o => o.SelectMany(t => t.tiles).Distinct());

        foreach(var owner in hasTerritory.Keys)
            result[owner] = hasTerritory[owner];

        return result;
    }
}

File: fsharp/go-counting/GoCounting.fs
module GoCounting

type Owner = | Black = 0 | White = 1 | None = 2

let map (board: string list) = 
    board |> Seq.indexed |> Seq.collect (fun (y, l) -> 
        l |> Seq.indexed |> Seq.map (fun (x, c) -> ((x,y),c)))
    |> Map.ofSeq

let findOwner tiles =
    tiles |> List.map snd |> List.except [' '] |> List.distinct 
        |> function
        | ['B'] -> Owner.Black
        | ['W'] -> Owner.White
        | _ -> Owner.None

let rec expand map path point =
    let c = Map.find point map
    match point with
    | p when c <> ' ' -> path @ [(p, c)]
    | (x,y) ->
        let newPath = path @ [(point, c)]
        let viable = 
            [(-1, 0);(1, 0);(0, -1);(0, 1)]
            |> List.map (fun (dx,dy) -> x + dx, y + dy)
            |> List.filter (fun p -> Map.containsKey p map && not (List.exists (fst >> (=) p) newPath))
        match viable with
        | [] -> newPath
        | _ -> viable |> List.collect (expand map newPath)

let extent (map: Map<(int * int), char> ) position =
    let tiles = expand map [] position |> List.distinct 
    let territory = tiles |> List.filter (snd >> (=) ' ') |> List.map fst |> List.sort
    (findOwner tiles, territory)

let territory board position = 
    let mapped = map board
    match mapped |> Map.tryFind position with
    | None -> None
    | Some c when c <> ' ' -> Some (Owner.None, [])
    | _ -> extent mapped position |> Some

let territories board =
    let mapped = map board
    mapped |> Map.toSeq |> Seq.map (fst >> extent mapped)
    |> Seq.append [(Owner.Black, []);(Owner.White, []);(Owner.None, [])] // ensures non-represented are part of final map
    |> Seq.groupBy fst |> Seq.map (fun (o,ps) -> 
        (o, ps |> Seq.collect snd |> Seq.distinct |> Seq.toList))
    |> Map.ofSeq